8335392: C2 MergeStores: enhanced pointer parsing

eme64 opened the pull request:
**Background**
I am introducing the `MemPointer`, for enhanced pointer parsing. For now, it replaces the much more limited `ArrayPointer` in `MergeStores` (see https://github.com/openjdk/jdk/pull/16245), but eventually it is supposed to be used widely in optimizations for pointer analysis: adjacency, aliasing, etc. I also plan to refactor the `VPointer` from auto-vectorization with it, and unlock more pointer patterns that way - possibly including scatter/gather.

**Details**

The `MemPointer` decomposes a pointer into the form `pointer = con + sum_i(scale_i * variable_i)` - a linear form with a sum of variables and scale-coefficients, plus some constant offset.

This form allows us to perform aliasing checks - basically we can check if two pointers are always at a constant offset. This allows us to answer many questions, including if two pointers are adjacent. `MergeStores` needs to know if two stores are adjacent, so that we can safely merge them.

More details can be found in the description in `mempointer.hpp`. Please read them when reviewing!

`MemPointer` is more powerful than the previous `ArrayPointer`: the latter only allows arrays, the former also allows native memory accesses, `Unsafe` and `MemorySegement`.

**What this change enables**

Before this change, we only allowed merging stores to arrays, where the store had to have the same type as the array element (`StoreB` on `byte[]`, `StoreI` on `int[]`).

Now we can do:
- Merging `Unsafe` stores to array. Including "mismatched size": e.g. `putChar` to `byte[]`.
- Merging `Unsafe` stores to native memory.
- Merging `MemorySegment`: with array, native, ByteBuffer backing types.
  - However: there is still some problem with RangeCheck smearing (a type of RC elimination) for the examples I have tried. Without RC's smeared, we can only ever merge 2 neighbouring stores. I hope we can improve this with better RangeCheck smearing. `MemorySegment` introduce `checkIndexL`, the long-variant of the RangeCheck. Normal array accesses only use the equivalent of `checkIndex`, the int-variant that we already optimize away much better.

**Dealing with Overflows**

We have to be very careful with overflows when dealing with pointers. For this, I introduced a `NoOverflowInt`. It allows us to do "normal" int operations on it, and tracks if there was ever an overflow. This way, we can do all overflow checks implicitly, and do not clutter the code with overflow-checks or - God forbid - forget overflow-checks.

**Benchmarks**

I added a few new benchmarks, to show the merging of `Unsafe` and `native` stores. We an see that 8 byte stores are now merged, and have the same performance as a long store. The same for 4 char stores that are merged into a single long store.

![image](https://github.com/user-attachments/assets/33b5cfcb-919b-46f4-bfa8-69fdff3acf1a)

And here the whole `MergeStores` benchmark, `master` and with `patch`:

![image](https://github.com/user-attachments/assets/16297539-a4b4-4593-a07e-b56eabb6952c)

All old benchmarks remain in the level of noise. A few have heavy errors - the benchmark ran on my laptop and maybe something slowed down for a bit. We can see the significant speedup with all the `unsafe` benchmarks at the bottom.

This is great, now we can perform `putChar` even on `byte[]`. This was an issue we discovered in this PR https://github.com/openjdk/jdk/pull/19626.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8335392](https://bugs.openjdk.org/browse/JDK-8335392): C2 MergeStores: enhanced pointer parsing (**Enhancement** - P4)


### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**)

### Contributors
 * Christian Hagedorn `<chagedorn@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19970/head:pull/19970` \
`$ git checkout pull/19970`

Update a local copy of the PR: \
`$ git checkout pull/19970` \
`$ git pull https://git.openjdk.org/jdk.git pull/19970/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19970`

View PR using the GUI difftool: \
`$ git pr show -t 19970`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19970.diff">https://git.openjdk.org/jdk/pull/19970.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19970#issuecomment-2419309510)
</details>

chhagedorn commented on src/hotspot/share/opto/mempointer.cpp:
```diff
@@ -0,0 +1,299 @@
+      {
+        // Decompose subtraction.
+        Node* a = n->in((opc == Op_AddP) ? 2 : 1);
+        Node* b = n->in((opc == Op_AddP) ? 3 : 2);
```
rm AddP

eme64 replied:
Good catch, will remove it here.

chhagedorn commented on src/hotspot/share/opto/mempointer.cpp:
```diff
@@ -0,0 +1,299 @@
+
+  // Compute distance:
+  NoOverflowInt distance = other.con() - con();
+  distance = distance.truncate_to_30_bits();
```
naming could be an issue, why coud it be NaN

eme64 replied:
Ok, I now replaced it with a `bool` method `is_abs_less_than_2_to_30`. I think that is more clear.

chhagedorn commented on src/hotspot/share/opto/traceMergeStoresTag.hpp:
```diff
@@ -0,0 +1,137 @@
+#include "utilities/bitMap.inline.hpp"
+#include "utilities/stringUtils.hpp"
+
+namespace TraceMergeStores {
```
can this be a class?

eme64 replied:
Not easily. Without the namespace, I get a clash with `traceAutoVectorization.hpp` names. A class could work, but then I need to initialize the arrays elsewhere - probably in a new cpp file. Annoying.

vnkozlov commented:
For me it is confusing to call `pointer = con + sum_i(scale_i * variable_i)` as "pointer" unless it is Unsafe address which has base address as constant.  It misses base address. All out pointer types are correspond to an address of some object in Java heap, out of heap, VM's object or some native (C heap) VM object.
This looks like `address_offset`, `displacement`, ...

vnkozlov commented on src/hotspot/share/opto/mempointer.hpp:
```diff
@@ -0,0 +1,487 @@
+//
+class MemPointerDecomposedForm : public StackObj {
+private:
+  // We limit the number of summands to 10. Usually, a pointer contains a base pointer
+  // (e.g. array pointer or null for native memory) and a few variables. For example:
+  //
+  //   array[j]                      ->  array_base + j + con              -> 2 summands
+  //   nativeMemorySegment.get(j)    ->  null + address + offset + j + con -> 3 summands
```
Does this means it supports only Unsafe and Array access?

eme64 replied:
I removed these examples, but instead added extensive examples at the beginning of `mempointer.hpp`

vnkozlov commented on src/hotspot/share/opto/mempointer.hpp:
```diff
@@ -0,0 +1,487 @@
+//
+//   Where each summand_i in summands has the form:
+//
+//     summand_i = scale_i * variable_i
```
I see you treat `scale` as compile time integer value (NoOverflowInt) and not Node. Why is the check/filter for that?

vnkozlov replied:
I mean: _Where_ is the check/filter for that?

eme64 replied:
The meat of the parsing happens in `MemPointerDecomposedFormParser::parse_sub_expression`.

One good example is this:
![image](https://github.com/user-attachments/assets/a83fd973-6407-45c6-8597-ef51edd7325d)

I added the `default` case now with a comment.

vnkozlov replied:
okay

eme64 commented:
@vnkozlov thanks for looking at this!

>For me it is confusing to call pointer = con + sum_i(scale_i * variable_i) as "pointer" unless it is Unsafe address which has base address as constant. It misses base address. All out pointer types are correspond to an address of some object in Java heap, out of heap, VM's object or some native (C heap) VM object.
This looks like address_offset, displacement, ...

I added some explanations and examples in the code now. But essencially, any `base` is just another `variable`, with `scale = 1`. Just for adjacency, it does not matter if the variable is some offset or a base address. Of course, there may be some other aliasing analysis tasks that do care if it is an array or not. We can add such detection later, if we need it.

eme64 commented:
I added this section to the desctiption above.

**Benchmarks**

I added a few new benchmarks, to show the merging of `Unsafe` and `native` stores. We an see that 8 byte stores are now merged, and have the same performance as a long store. The same for 4 char stores that are merged into a single long store.

![image](https://github.com/user-attachments/assets/33b5cfcb-919b-46f4-bfa8-69fdff3acf1a)

eme64 commented:
I added this to the desciption:

**What this change enables**

Before this change, we only allowed merging stores to arrays, where the store had to have the same type as the array element (`StoreB` on `byte[]`, `StoreI` on `int[]`).

Now we can do:

Merging `Unsafe` stores to array. Including "mismatched size": e.g. `putChar` to `byte[]`.
Merging `Unsafe` stores to native memory.
Merging `MemorySegment`: with array, native, ByteBuffer backing types.

vnkozlov commented on src/hotspot/share/opto/mempointer.hpp:
```diff
@@ -0,0 +1,591 @@
+//   The con and scale_i are compile-time constants (NoOverflowInt), and the variable_i are
+//   compile-time variables (C2 nodes).
+//
+//   For the MemPointer, we do not explicitly track base address. For Java heap pointers, the
+//   base address is just a variable. For native memory (C heap) pointers, the base address is
```
May be: "For Java heap pointers, the  base address is just a variable  in summand with `scale` == 1.

vnkozlov commented on src/hotspot/share/opto/mempointer.hpp:
```diff
@@ -0,0 +1,591 @@
+// Intuition and Examples:
+//   We parse / decompose pointers into a linear form:
+//
+//     pointer = con + sum_i(scale_i * variable_i)
```
May be you can swap them `sum_i(variable_i * scale_i) + con`

vnkozlov reviewed (commented):
This is more clear now.

eme64 commented on src/hotspot/share/opto/memnode.cpp:
```diff
@@ -3051,62 +2904,47 @@ StoreNode* MergePrimitiveArrayStores::run() {
   assert(opc == Op_StoreB || opc == Op_StoreC || opc == Op_StoreI, "precondition");
-  assert(_store->adr_type()->isa_aryptr() != nullptr, "must be array store");
-  assert(!_store->is_unsafe_access(), "no unsafe accesses");
+  // assert(_store->adr_type()->isa_aryptr() != nullptr, "must be array store");
```
remove!

vnkozlov reviewed (approved):
Looks good to me.

dean-long commented on src/hotspot/share/opto/noOverflowInt.hpp:
```diff
@@ -0,0 +1,115 @@
+  // Create from jlong (or jint) -> NaN if overflows jint.
+  explicit NoOverflowInt(jlong value) : _is_NaN(true), _value(0) {
+    jint trunc = (jint)value;
+    if ((jlong)trunc == value) {
```
Do you think we need these runtime checks even in product builds?  If not, then consider using checked_cast<>() here.

eme64 replied:
I definitely need these checks in product, yes. I want longs outside the int-range to become NaN.

dean-long commented on src/hotspot/share/opto/noOverflowInt.hpp:
```diff
@@ -0,0 +1,115 @@
+  friend NoOverflowInt operator-(const NoOverflowInt a, const NoOverflowInt b) {
+    if (a.is_NaN()) { return make_NaN(); }
+    if (b.is_NaN()) { return make_NaN(); }
+    return NoOverflowInt(java_subtract((jlong)a.value(), (jlong)b.value()));
```
For add and subtract, I don't think two 32-bit inputs can cause a 64-bit result to wrap or overflow, but it could then using the java_ APIs that wrap would hide the overflow.  I think you can safely use + and - for these two.

eme64 replied:
Ok, I'll change it to `jlong +/-`

dean-long commented on src/hotspot/share/opto/noOverflowInt.hpp:
```diff
@@ -0,0 +1,115 @@
+  friend NoOverflowInt operator*(const NoOverflowInt a, const NoOverflowInt b) {
+    if (a.is_NaN()) { return make_NaN(); }
+    if (b.is_NaN()) { return make_NaN(); }
+    return NoOverflowInt(java_multiply((jlong)a.value(), (jlong)b.value()));
```
```suggestion
    return NoOverflowInt((jlong)a.value() * (jlong)b.value());
```
Can't overflow 64-bits, so no need for java_multiply

dean-long commented on src/hotspot/share/opto/noOverflowInt.hpp:
```diff
@@ -0,0 +1,115 @@
+    if (b.is_NaN()) { return make_NaN(); }
+    jint shift = b.value();
+    if (shift < 0 || shift > 31) { return make_NaN(); }
+    return NoOverflowInt(java_shift_left((jlong)a.value(), shift));
```
```suggestion
    return NoOverflowInt((jlong)a.value() << shift);
```
left shift for shift values between 0 and 31 should be well-defined.

dean-long commented on src/hotspot/share/opto/noOverflowInt.hpp:
```diff
@@ -0,0 +1,115 @@
+  }
+
+  NoOverflowInt abs() const {
+    if (is_NaN()) { return make_NaN(); }
```
```suggestion
    if (is_NaN() || value() == min_jint) { return make_NaN(); }
```

eme64 replied:
It is not necessary, and I actually test that in my gtest. The subtraction below handles the `min_jint` case.

eme64 replied:
`ASSERT_TRUE(NoOverflowInt(min_jint).abs().is_NaN());`

But if you insist on it I can add it in.

dean-long commented on src/hotspot/share/opto/noOverflowInt.hpp:
```diff
@@ -0,0 +1,115 @@
+  }
+
+  friend NoOverflowInt operator<<(const NoOverflowInt a, const NoOverflowInt b) {
+    if (a.is_NaN()) { return make_NaN(); }
+    if (b.is_NaN()) { return make_NaN(); }
```
```suggestion
    if (a.is_NaN()) { return a; }
    if (b.is_NaN()) { return b; }
```
This might be more efficient than creating a new one.

dean-long commented on src/hotspot/share/opto/noOverflowInt.hpp:
```diff
@@ -0,0 +1,115 @@
+    }
+  }
+
+  static NoOverflowInt make_NaN() { return NoOverflowInt(); }
```
```suggestion
  static constexpr NoOverflowInt make_NaN() { return NoOverflowInt(); }
```
I think this can be constexpr, but you may need to add constexpr to the ctor as well.

eme64 commented:
@vnkozlov thanks for the review!
@dean-long thanks for the comments. I addressed all your comments.

dean-long commented:
OverflowInt updates look good.  (That's the only part I reviewed.)

eme64 commented:
@dean-long ok, thanks for the NoOverflowInt review :blush: 

In that case, I think we need yet another reviewer @vnkozlov , right?

vnkozlov reviewed (approved):
Latest code is good.

dean-long commented:
BTW, another place where a user-defined class like NoOverflowInt might be useful is compiler invocation counters.  We store them as integers in various places, increment and scale them, check for saturation, and convert them to floating point for comparison.  The last time I tried to fix a bug in that area, I was wishing for a better way to propagate integer saturation through multiple operations (like how NoOverflowInt uses NaN), and was tempted to change everything to floating point so I could use infinity.

chhagedorn reviewed (commented):
Nice work! I have a first round of comments - mostly minor things. So far, it looks good. Will pick this up again tomorrow.

eme64 commented on src/hotspot/share/opto/mempointer.hpp:
```diff
@@ -0,0 +1,591 @@
+// -----------------------------------------------------------------------------------------
+//
+// MemPointerDecomposedForm:
+//   When the pointer is parsed, it is decomposed into sum of summands plus a constant:
```
obsolete after your other suggestion

eme64 commented on src/hotspot/share/opto/mempointer.hpp:
```diff
@@ -0,0 +1,591 @@
+//     MemorySegment ms = Arena.ofAuto().allocate(1000, 1);
+//     assert ms.heapBase().isEmpty(): "null base";
+//     assert ms.address() != 0: "non-zero native memory address";
+//     byte val2 = ms.get(ValueLayout.JAVA_BYTE, i);
+//
+//     pointer = ms.heapBase() +           ms.address() +           i
+//             = 0             + 1       * ms.address() + 1       * i
```
good idea!

eme64 commented on src/hotspot/share/opto/mempointer.hpp:
```diff
@@ -0,0 +1,591 @@
+    _aliasing(aliasing),
+    _distance(distance)
+  {
+    const jint max_distance = 1 << 30;
+    assert(_distance < max_distance && _distance > -max_distance, "safe distance");
```
Good point. I'll try to remember the reason and add better comments.

eme64 replied:
Woopsies. I think this was a left-over from something earlier. I was able to trigger this assert with this:

```
    static long[] arr201 = new long[1 << 28];

    public static void test201() {
        UNSAFE.putByte(arr201, Unsafe.ARRAY_LONG_BASE_OFFSET + (1L << 12),              (byte)64);
        UNSAFE.putByte(arr201, Unsafe.ARRAY_LONG_BASE_OFFSET + (1L << 12) + (1L << 30), (byte)64);
    }
```

eme64 replied:
Good catch! Thanks for having a close look!

eme64 replied:
I added some regression tests for that case, and fixed the assert.

eme64 commented:
@chhagedorn thanks for all your comments. I addressed all now.
I had to fix that one assert in `MemPointerAliasing` constructor - it was wrong and I added regression tests for it now :blush: 

Looking forward to the next part of your review!

eme64 commented:
FYI: I ran performance testing, with no significant changes detected. That was expected (this is a niche optimization, that probably does not feature prominently in the benchmarks - especially Unsafe stores are not that prevalent). Still, this change is justified by my added micro-benchmarks.

wenshao commented:
In the toString scenario of Integer/Long and the StringBuilder.appendNull/appendBoolean scenario, we can refactor the code to optimize based on unsafe mergestore. I am waiting for this PR to be merged, and then continue to complete PR 19626.

eme64 commented:
@wenshao Thanks for your patience. @chhagedorn is doing a thorough review right now, so I hope we are only a few days away from integration ;)

eme64 commented on src/hotspot/share/opto/mempointer.cpp:
```diff
@@ -0,0 +1,379 @@
+  if (_con.is_NaN()) { return MemPointerDecomposedForm(pointer); }
+
+  // Sort summands by variable->_idx
+  _summands.sort(MemPointerSummand::cmp_for_sort);
```
good idea!

eme64 commented on src/hotspot/share/opto/noOverflowInt.hpp:
```diff
@@ -0,0 +1,115 @@
+  jint value() const { assert(!is_NaN(), "NaN not allowed"); return _value; }
+  bool is_zero() const { return !is_NaN() && value() == 0; }
+
+  friend NoOverflowInt operator+(const NoOverflowInt a, const NoOverflowInt b) {
```
Good idea!

eme64 commented on src/hotspot/share/opto/noOverflowInt.hpp:
```diff
@@ -0,0 +1,115 @@
+  }
+
+  NoOverflowInt abs() const {
+    if (is_NaN()) { return make_NaN(); }
```
Yes, I changed it!

eme64 commented on src/hotspot/share/opto/mempointer.hpp:
```diff
@@ -0,0 +1,612 @@
+    assert(_distance != min_jint, "given by condition S3 of MemPointer Lemma");
+  }
+
+public:
+  MemPointerAliasing() : MemPointerAliasing(Unknown, 0) {}
```
I just removed this constructor!

eme64 commented on src/hotspot/share/opto/mempointer.hpp:
```diff
@@ -0,0 +1,612 @@
+public:
+  MemPointerAliasing() : MemPointerAliasing(Unknown, 0) {}
+
+  static MemPointerAliasing make_unknown() {
+    return MemPointerAliasing();
+  }
```
Yes, this seems better, I'm doing this :)

eme64 commented on src/hotspot/share/opto/mempointer.hpp:
```diff
@@ -0,0 +1,612 @@
+  MemPointerSummand() :
+      _variable(nullptr),
+      _scale(NoOverflowInt::make_NaN()) {}
+  MemPointerSummand(Node* variable, const NoOverflowInt scale) :
```
Will do that, and similarly elsewhere!

eme64 commented on src/hotspot/share/opto/mempointer.hpp:
```diff
@@ -0,0 +1,612 @@
+//
+class MemPointerDecomposedForm : public StackObj {
+private:
+  // We limit the number of summands to 10. Usually, a pointer contains a base pointer
+  // (e.g. array pointer or null for native memory) and a few variables.
+  static const int SUMMANDS_SIZE = 10;
```
Ok, will do

eme64 commented on src/hotspot/share/opto/mempointer.hpp:
```diff
@@ -0,0 +1,612 @@
+  }
+
+private:
+  MemPointerDecomposedForm(Node* pointer, const GrowableArray<MemPointerSummand>& summands, const NoOverflowInt con)
```
did that

chhagedorn commented on src/hotspot/share/opto/mempointer.cpp:
```diff
@@ -0,0 +1,382 @@
+  }
+#endif
+
+  // "MemPointer Lemma" condition S2: check if all summands are the same:
```
```suggestion
  // "MemPointer Lemma" condition (S2): check if all summands are the same:
```

chhagedorn commented on src/hotspot/share/opto/mempointer.cpp:
```diff
@@ -0,0 +1,382 @@
+    }
+  }
+
+  // "MemPointer Lemma" condition S3: check that the constants do not differ too much:
```
```suggestion
  // "MemPointer Lemma" condition (S3): check that the constants do not differ too much:
```

chhagedorn commented on src/hotspot/share/opto/mempointer.cpp:
```diff
@@ -0,0 +1,382 @@
+    return MemPointerAliasing::make_unknown();
+  }
+
+  // "MemPointer Lemma" condition S1:
```
```suggestion
  // "MemPointer Lemma" condition (S1):
```

chhagedorn commented on src/hotspot/share/opto/mempointer.cpp:
```diff
@@ -0,0 +1,382 @@
+  //   same memory object. With the Pre-Condition, we know that both pointers are in
+  //   bounds of that same memory object.
+
+  // Hence, all 3 conditions of the "MemoryPointer Lemma" are established, and hence
```
Since we also have added `(S0)` recently, we might need to add a word here about it and then update this to "all 4 conditions".

eme64 replied:
Good idea

chhagedorn commented on src/hotspot/share/opto/mempointer.cpp:
```diff
@@ -0,0 +1,382 @@
+
+  return is_adjacent;
+}
+
```
Two new lines:
```suggestion
```

chhagedorn reviewed (commented):
Some final last mostly minor comments but otherwise, it looks good to me now! I like the summaries and how you worked out the proofs. They are now easy to understand.

eme64 commented:
/contributor add chhagedorn

You spent enough time on this already ;)

eme64 commented on src/hotspot/share/opto/memnode.cpp:
```diff
@@ -3051,62 +2904,46 @@ StoreNode* MergePrimitiveArrayStores::run() {
+  const MemPointer pointer_use(use_store NOT_PRODUCT( COMMA trace ));
+  const MemPointer pointer_def(def_store NOT_PRODUCT( COMMA trace ));
+  if (!pointer_def.is_adjacent_to_and_before(pointer_use)) {
     return false;
   }
-
   return true;
```
Ah good idea

chhagedorn commented:
> /contributor add chhagedorn
> 
> You spent enough time on this already ;)

Thanks Emanuel, I highly appreciate that :-)

eme64 commented on src/hotspot/share/opto/mempointer.cpp:
```diff
@@ -0,0 +1,382 @@
+  int traversal_count = 0;
+  while (_worklist.is_nonempty()) {
+    // Bail out if the graph is too complex.
+    if (traversal_count++ > 1000) { return MemPointerDecomposedForm(pointer); }
```
Good idea!

eme64 commented on test/hotspot/jtreg/compiler/c2/TestMergeStoresMemorySegment.java:
```diff
@@ -0,0 +1,426 @@
+ * @run driver compiler.c2.TestMergeStoresMemorySegment Native
+ */
+
+// FAILS: mixed providers currently do not merge stores. Maybe there is some inlining issue.
```
https://bugs.openjdk.org/browse/JDK-8343536

eme64 commented on test/hotspot/jtreg/compiler/c2/TestMergeStoresMemorySegment.java:
```diff
@@ -0,0 +1,426 @@
+        phase = CompilePhase.PRINT_IDEAL,
+        applyIf = {"UseUnalignedAccesses", "true"})
+    static Object[] test_xxx(MemorySegment a, int xI, int yI, int zI) {
+         // All RangeChecks remain -> RC smearing not good enough?
```
https://bugs.openjdk.org/browse/JDK-8343536

eme64 commented:
@chhagedorn I filed https://bugs.openjdk.org/browse/JDK-8343536 to track the cases in `TestMergeStoresMemorySegment.java` that do not optimize as hoped for.

eme64 commented:
@chhagedorn I addressed all your review suggestions. Thank you very much for the in-depth review :)

eme64 commented:
@vnkozlov Would you like to re-review? If I don't hear anything then I'll integrate tomorrow.

vnkozlov reviewed (approved):
Update is good.

eme64 commented:
Thanks @chhagedorn for the extensive reviews and collaboration on improving the proofs :blush: 
Thanks @vnkozlov for the approval.

I did an offline merge and testing (to avoid requiring a re-approval) - all looks good.

/integrate

wenshao commented:
@eme64 How do I use the TraceMergeStores option? It worked before, but now it gives an error.

```sh
build/macosx-aarch64-server-fastdebug/jdk/bin/java -Dtest=appendNullLatin1 -XX:+TraceMergeStores
```

output
```
Unrecognized VM option 'TraceMergeStores'
Did you mean '(+/-)MergeStores'?
Error: Could not create the Java Virtual Machine.
Error: A fatal exception has occurred. Program will exit.
```

eme64 commented:
@wenshao Ah, good question! I changed it from a "global" flag to a "compile option". You can now filter the methods! And you can enable different tags - so you can regulate how verbose it is. Example:
`-XX:CompileCommand=TraceMergeStores,Test::test*,SUCCESS,ADJACENCY,ALIASING,BASIC`

And to see all available tags:
`-XX:CompileCommand=TraceMergeStores,Test::test*,help`

```
Usage for CompileCommand TraceMergeStores:
  -XX:CompileCommand=TraceMergeStores,<package.class::method>,<tags>
  tags                   descriptions
  BASIC                  Trace basic analysis steps
  POINTER                Trace pointer IR
  ALIASING               Trace MemPointerSimpleForm::get_aliasing_with
  ADJACENCY              Trace adjacency
  SUCCESS                Trace successful merges
```

You might have to play around a little to see what is helpful to you. And I'm always open to feedback :)

wenshao commented:
Currently, TraceMergeStores can only be used in fastdebug images. Are you planning to support it in release images?

eme64 commented:
> Currently, TraceMergeStores can only be used in fastdebug images. Are you planning to support it in release images?

I generally don't support it in release builds, only debug - or rather `NOT_PRODUCT`. The issue with supporting in release is that some other printing methods I use are not available in product (`Node::dump`). And if we support it in release, then we have to create a CSR, clearly specify what it prints, and then we are going to be less flexible in the future with changing the behavior. So I would rather not make it product, at least for now ;)
BTW: this is also why you can only disable `-XX:-MergeStores` with `-XX:+UnlockDiagnosticVMOptions `: it is not a full product flag, and so does not require a CSR, and we are able to remove it or change its behavior. But if someone really has an issue with the MergeStores optimization, they at least have a workaround until we are able to fix it ;)

Does that make sense?

wenshao commented:
If it is not provided in the release image, users need to find the source code of the current version of JDK to build the fastdebug image to analyze whether the MergeStore optimization of a certain code works.

I can understand that MergeStore may still need to be improved, so it cannot be used as a product feature, but this is a useful optimization and I hope it can be provided in the product eventually.

I hope that TraceMergeStore can eventually be used in the release image like `PrintInlining` and become a tool for performance optimizers.

eme64 commented:
@wenshao I suppose we could consider making `TraceMergeStores` and `TraceAutoVectorization` available in product, but under the `-XX:+UnlockDiagnosticVMOptions` flag... I will discuss this with other VM engineers. That means it is available, but there is no promise of stability. Still, once people become dependent on it, maybe even tools become dependent, then it is harder to make changes without everybody complaining :sweat_smile:
